File: python/construct_bst.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, value):
        node = Node(value)
        if not self.root:
          self.root = node
          return
        tree = self.root
        while True:
          if value< tree.value:
            #move to left
            if not tree.left:
              tree.left = node
              return self
            tree = tree.left
          else:
            #move to right
            if not tree.right:
              tree.right = node
              return self
            tree = tree.right

    def find(self, value):
        if not self.root: return False
        tree = self.root
        while tree:
          if value== tree.value:
            return tree
          elif value< tree.value:
            tree = tree.left
          else:
            tree = tree.right
        return False

    def remove(self, value, current=None, parent=None):
        if not self.root: return False
        current = self.root
        while current:
          if value< current.value:
            parent = current
            current = current.left
          elif value> current.value:
            parent = current
            current = current.left
          else:
            #found the node to be deleted
            #node to be deleted has 2 childred
            if current.left and current.right:
                current.value = self.get_min(current.right)
                self.remove(current.value, current.right, current)
            #if deleting the root node
            elif parent is None:
                if current.left:
                    current.value = current.left.value
                    current.right = current.left.right
                    current.left = current.left.left
                elif current.right:
                    current.value = current.right.value
                    current.right = current.right.right
                    current.left = current.right.left
                else:
                    #//this is a single node bst
                    self.root = None
            elif current == parent.left:
                parent.left = current.left if current.left else current.right
            elif current == parent.right:
                parent.right = current.left if current.left else current.right
            break
            
        return self

    def get_min(self, node):
        while node.left:
            node = node.left
        return node.value

    

File: javascript/construct_bst.js
class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

class BinarySearchTree {
    constructor() {
        this.root = null;
    }
    insert(value) {
        const node = new Node(value);
        if (!this.root) {
            this.root = node;
            return this;
        }
        let tree = this.root;
        while (true) {
            if (value < tree.value) {
                //move left
                if (!tree.left) {
                    tree.left = node;
                    return this;
                }
                tree = tree.left;
            } else {
                //move right
                //value > = tree.value
                if (!tree.right) {
                    tree.right = node;
                    return this;
                }
                tree = tree.right;
            }
        }
    }

    find(value) {
        if (!this.root) return false;
        let tree = this.root;
        while (tree) {
            if (value < tree.value) tree = tree.left;
            else if (value > tree.value) tree = tree.right;
            else if (value === tree.value) return tree;
        }
        return false;
    }

    remove(value, current = this.root, parent = null) {
        if (!this.root) return false;
        while (current) {
            if (value < current.value) {
                parent = current;
                current = current.left;
            }
            else if (value > current.value) {
                parent = current;
                current = current.right;
            }
            else {
                //found the node to be deleted
                //node to be deleted has 2 childred
                if (current.left !== null && current.right !== null) {
                    current.value = this.getMin(current.right);
                    this.remove(current.value, current.right, current);
                }
                // if deleting the root node
                else if (parent === null) {
                    if (current.left !== null) {            // Incase root doesn't have right clild
                        current.value = current.left.value;
                        current.left = current.left.left;
                        current.right = current.left.right;
                    } else if (current.right !== null) {    // Incase root doesn't have right clild
                        current.value = current.right.value;
                        current.left = current.right.left;
                        current.right = current.right.right;
                    } else {
                        //this is a single node bst
                        this.root = null;
                    }
                }
                else if (current === parent.left) {
                    parent.left = current.left !== null ? current.left : current.right;
                }
                else if (current === parent.right) {
                    parent.right = current.left !== null ? current.left : current.right;
                }
                //break out of the while loop
                break;
            }
        }
        return this;
    }

    getMin(node) {
        while (node.left !== null) {
            node = node.left;
        }
        return node.value;
    }
}
//                     20
//                    /   \
//                    6    35
//                   / \   /  \ 
//                   3  8  27  55
//                  / \    / \   \
//                  1  3   25 29  60

let bst = new BinarySearchTree();


bst.insert(20);
bst.insert(6);
bst.insert(35);
bst.insert(3);
bst.insert(8);
bst.insert(27);
bst.insert(55);
bst.insert(1);
bst.insert(3);
bst.insert(25);
bst.insert(29);
bst.insert(60);
bst.insert(70)
console.log(bst.find(27))
console.log(bst.remove(55));
